[WIP] add `Request.retry`

popeindustries opened the pull request:
I'm just opening this early sketch to see if there is interest in adding a `Request.retry` method. I'm aware of [segmentio/superagent-retry](https://github.com/segmentio/superagent-retry), but it relies on some hacky behaviour to reset state between attempts, and doesn't seem to work with newer versions of `superagent`. There would seem to be quite a bit of resetting necessary to make this work properly for errors, timeouts, and aborts, so I think it might be better done as a core method rather than a plugin.

In any case, I'd be happy to work on adding this if there is interest.

Progress:
- [x] Node: retry for errors
- [x] Node: retry for timeouts
- [x] Browser: retry for errors
- [x] Browser: retry for timeouts
- [x] Node: retry with redirects
- [x] refactor `end()` into 2 phases for "lightweight" retry?

pornel commented:
I think that would be a useful addition.

As you've noted, superagent has a quite a bit of internal state. Do you have a plan how to deal with that nicely?

popeindustries commented:
"Nicely" will definitely be the challenge ;)

I don't yet have a full picture, but it looks like `callback()` would be the best place to trigger a reset (`_aborted`, `timedout`, `called`, `_endCalled`, etc) and re-execution of `end()`. I think the real challenge will be dealing with the event handlers for the internal `req` instance. On timeout, for instance, after `abort()` is called and an `err` passed to `callback`, if we reset `_aborted`, the error handler [here](https://github.com/visionmedia/superagent/blob/master/lib/node/index.js#L597) will not return early and we'd end up with a double callback.

pornel commented:
I suspect it may be necessary to change `.end()` and split it into 2 phases: things done once only (e..g append query string), and things done to launch the request (the latter call being repeated when retrying).

popeindustries commented:
Ok, thanks. I'll start with some tests and start poking around.

popeindustries commented:
@pornel I think things are looking pretty good so far. I just have a couple questions if you have time to look things over:

- do you have any suggestions on how to avoid [this](https://github.com/popeindustries/superagent/blob/67c970fd78a4980f9caf5adf36433c03b6bd1075/lib/node/index.js#L641-L649) mess when resetting `_timeout` or `_responseTimeout` after they have been cleared in `clearTimeout`?
- is it necessary to reset headers [here](https://github.com/popeindustries/superagent/blob/67c970fd78a4980f9caf5adf36433c03b6bd1075/lib/request-base.js#L160)?

I've so far avoided touching `end()`. I can see how it would be more efficient to avoid initial parsing on each retry, but I'm a little hesitant to mess with anything in there if it's not absolutely necessary.

pornel commented on lib/client.js:
```diff
@@ -590,10 +593,25 @@ Request.prototype._getFormData = function(){
+    var timeout = 0;
+    var responseTimeout = 0;
+    if (err && err.timeout) {
+      if (err.message.indexOf('Response') == 0) {
```
That looks odd.

1. You shouldn't need to read the timeout from the error. Read request's properties set by the `timeout()` function directly
2. *if* you needed to identify error thrown by superagent, add a property to it, don't touch the message

popeindustries replied:
Yeah, this is really ugly because `_timeout` and `_responseTimeout` are zeroed out in `clearTimeout`, before `callback` is called. Any reason we can't just delete [L45-L46](https://github.com/visionmedia/superagent/blob/master/lib/request-base.js#L45-L46) in `request-base`?

pornel replied:
Oh, yes, delete it there.

pornel commented on lib/client.js:
```diff
@@ -590,10 +593,25 @@ Request.prototype._getFormData = function(){
   this.clearTimeout();
 
   if (err) {
+    if (this._maxRetries) err.attempts = this._retries;
```
Maybe `err.retries`?

pornel commented on lib/node/index.js:
```diff
@@ -595,6 +598,8 @@ Request.prototype.request = function(){
     // when request is aborted before a connection is made
     if (self._aborted) return;
+    // flag retry here for out timeouts (same as above)
+    if (self._retries) return;
```
I don't understand why it skips errors here. 

Also note that all these handlers are `.once()`, so they should have been re-attached again on retry. That's why I've mentioned earlier that .end() may need to be split up into once-ever and once-per-retry functions.

popeindustries replied:
This is also a little strange. After aborting `req` (timeout -> `abort()`), I delete the **old** `req`, but if I also `req.removeAllListeners()`, an unhandled "socket hang up" error is thrown from the **old** instance. If I don't remove listeners on the **old** `req` instance, then I have to skip somehow. 

I think the best approach would be to memoize the current retry index and return early if they are not the same.

pornel commented on lib/request-base.js:
```diff
@@ -130,6 +130,50 @@ RequestBase.prototype.timeout = function timeout(options){
+
+RequestBase.prototype.retry = function retry(count){
+  this._maxRetries = count || 1;
+  this._retries = 0;
```
That is set in `end()`, so you shouldn't need to change that

popeindustries replied:
I'm not sure I follow. This is the main api for enabling `retry` (like `timeout`). I tried to follow the same logic as for redirects.

pornel replied:
Sorry, I've meant it's set in https://github.com/visionmedia/superagent/pull/1155/files#diff-c24ce7e3da4c0e4ff811a2b6a76f8bd9R145

It's not an error, but it's redundant. `_retries` are never increased before `.retry()` is called, so there's no point zeroing something that's zero anyway.

pornel commented on lib/request-base.js:
```diff
@@ -130,6 +130,50 @@ RequestBase.prototype.timeout = function timeout(options){
+  delete this._responseTimeoutTimer;
+
+  this._endCalled = false;
+  this.end(this._callback);
```
Doesn't this break things? e.g. append query string twice?

popeindustries replied:
Yes, I can see how this would be a problem. I'll take a look at refactoring `end()` next.

pornel commented on lib/should-retry.js:
```diff
@@ -0,0 +1,22 @@
+  if (err && err.code && ~ERROR_CODES.indexOf(err.code)) return true;
+  if (res && res.status && res.status >= 500) return true;
+  // Superagent timeout
+  if (err && 'timeout' in err && err.code == 'ECONNABORTED') return true;
```
Please make sure that `.abort()` called by the user stops all retries. This may require creating a private abort for superagent's timeouts.

popeindustries replied:
Good point. I'll write a test for this

popeindustries replied:
I can confirm that this works fine (via test). Calling `abort()` never triggers `callback()`, so no more retries are attempted.

pornel commented on test/support/server.js:
```diff
@@ -417,4 +417,27 @@ app.get('/if-mod', function(req, res){
 });
 
+var errAttempts = 0;
+app.get('/error/ok', function(req, res) {
```
In CI we have multiple browsers running the test at the same time, and I think they're sharing the server, so they could interfere with each other's attempts.

It may be more robust to take a random ID from the client as an argument, and count retries per that ID.

popeindustries replied:
Of course! I was wondering why the tests worked locally but not remotely.

pornel commented:
Answering your questions:

1. If you split and reuse part of `end()` that sets these timeouts, you won't need it

2. Doesn't seem needed. Does it break any tests?

3. I'd use `retries`, but `attempts` is fine too.

pornel commented:
> I'm a little hesitant to mess with anything in there if it's not absolutely necessary.

I could try to do that refactoring first if you want

popeindustries commented:
I've split `end()` into public/private methods as you suggested. Seems to work fine. I'm not sure if there is anything else I've forgotten, but the browser tests are still failing. 

I've added an (pseudo) unique id to the requests to work around the parallelisation problem you raised, but there still seems to be something wrong with the tests that require different responses from the same endpoint.

pornel commented on lib/request-base.js:
```diff
@@ -155,17 +155,16 @@ RequestBase.prototype._retry = function() {
 
   this._aborted = false;
   this.timedout = false;
   delete this._timer;
```
I don't understand why these fields are deleted here. The timers are probably cancelled when the error happens (they should…), and should be irrelevant later.

popeindustries replied:
In `_setTimeouts()` there is the following check, so if they are not removed, the new timers for retry can't be set:
```js
if (this._timeout && !this._timer) {
```
Probably better to delete or zero out in `clearTimeout()`?

pornel commented on lib/request-base.js:
```diff
@@ -155,17 +155,16 @@ RequestBase.prototype._retry = function() {
-  this._endCalled = false;
-  this.end(this._callback);
-  return this;
+  return this._end();
```
That looks good!

pornel commented:
Thank you. 

The failures were only due to short timeouts in the test.

popeindustries commented:
That's great! Thanks for your help and patience
